spec/sidekiq_helper.rb

```diff
@@ -0,0 +1,53 @@
+# frozen_string_literal: true
+
+PID_FILE = File.join(Rails.root, 'tmp', 'pids', 'sidekiq.pid')
+LOG_FILE = File.join(Rails.root, 'tmp', 'sidekiq-log.log')
+
+def read_pid
+  return nil unless File.exists? PID_FILE
+  File.open(PID_FILE).read.strip
+end
+
+def write_pid(pid)
+  File.open(PID_FILE, 'w') {|f| f.print pid }
+end
+
+def sidekiq_running?
+  begin
+    !!Process.getpgid(read_pid)
+  rescue
+    false
+  end
+end
+
+def start_sidekiq
+  unless sidekiq_running?
+    write_pid(spawn("bundle exec sidekiq -L #{LOG_FILE}"))
+  end
+end
+
+def stop_sidekiq
+  Process.kill(HUP, read_pid) if sidekiq_running?
+  File.open(LOG_FILE, 'w') {}
+end
+
+def sidekiq_log
+  File.read(LOG_FILE)
+end
+
+RSpec.configure do |config|
+  config.before(:each) do |example|
+    if example.metadata[:sidekiq]
+      start_sidekiq
```

Comment:
I am not keen on this. I would prefer that we make it a dependency of running the test suite that Sidekiq already be running in the test environment (with `RAILS_TEST bundle exec sidekiq`), and just:
1. Document that in the README
2. Run `bundle exec sidekiq` in Travis's `before_script`

Reply:
What is your concern here?
The only two downsides I see with the current approach are:
1) There's a danger of the process not starting/stopping correctly(which it wasn't, but while confirming what I've written here is accurate I've made sure the logic is more reliable.) However this same risk exists for starting/stopping the sidekiq process manually, with the only difference being that it's more likely to be spotted by the person initiating/terminating the process when they're looking at it in a distinct terminal.
2) Having to include the 'sidekiq: true' flag on specs that make use of this functionality.

The upsides on the other hand are:
1) Less overhead to running tests in the develop environment, especially considering we already need to run docker, so just to run tests you'd need to have three terminals open(or start and keep track of daemons). One for docker, one for sidekiq and a third for rspec to run in.
2) Counter to point 2. above, because we only start up sidekiq on specs with the flag, when running subsets of tests we don't necessarily need and won't start sidekiq.
3) Because sidekiq is started with the same codebase and env settings as the rspec task, it will keep settings in sync between the test 'worker' and 'web' threads. So we don't need to worry about restarting sidekiq manually just because code, test settings or gems were updated. This is especially useful for code changes to models that are used within workers or the jobs themselves.
4) Having rspec start start the sidekiq process seems better in the sense that it gives more control to the test environment as opposed to relying on external factors. So tests should be more consistent, since the sidekiq process will always be started in exactly the same way.

Reply:
We should not be attempting to replicate the functionality of `bundle exec sidekiq`, which is what we use to run Sidekiq in production. In a *system* test, we should run exactly that to test that *it* works, not that something else does.

I suggest that we:
1. Reduce spec/sidekiq_helper.rb to:
    ```ruby
    # frozen_string_literal: true

    RSpec.configure do |config|
      config.before(:each) do |example|
        if example.metadata[:sidekiq]
          Sidekiq::Testing.disable!
        end
      end

      config.after(:each) do |example|
        Sidekiq::Testing.fake! if example.metadata[:sidekiq]
      end
    end
    ```
2. In spec/system/story_submittal_and_retrieval_spec.rb, at line 40, add:
    ```ruby
    # Sidekiq needs to be running to continue with this system spec
    expect(Sidekiq::ProcessSet.new.size).to be_nonzero, 'Sidekiq is not running. Run it with: `RAILS_ENV=test bundle exec sidekiq`'
    ```

This means that, yes, to have all tests pass, you will need to have pre-run `RAILS_ENV=test bundle exec sidekiq`, but if you are just running some unit tests, and not the full suite, you need not.

Reply:
I disagree, since sidekiq is what runs code part of the system. 
If we wanted to only test our codebase, then we should be using the sidekiq testing helpers and running jobs inline with `Sidekiq::Testing.inline!`. 

However I think it's valid to want to try to not only make sure that our code works, but also that our code works when running via sidekiq.

In order to do that though we need to keep the code running in the sidekiq process at the same version as the code being tested, otherwise we're not testing anything that would ever be actually deployed anywhere. 

For example if we changed the parameters that our thumbnail job takes and added a new parameter, tests would fail because sidekiq wouldn't be able to pick them up(until a developer restarted it). On the other hand though, if we removed a parameter from the job there's potential for the job to still work on the old sidekiq processor, as long as the previous versions of the jobs handle nils and satisfy the expectations. Then when the new version of the code is running in sidekiq(after a developer has manually restarted it while the new version of the code is checked out) the tests may fail because of the changes to the job(s).

The only use case of being able to run different versions of our workers to the corresponding application code would be testing new deployments, in any other situation it is undesirable to have the two things out of sync.

So, if we indeed wanted to take this approach of leaving the starting of the job processing to the developer that is running tests, then the check for sidekiq would not only need to ensure it's responding, but also that it is running the correct version of the codebase. I'm not sure if there's a way to inspect/infer this.

Further, with the current helper, I don't think we're _replicating_ the functionality of `bundle exec sidekiq`, rather we're just ensuring that it _is_ running. I think letting rspec start sidekiq is a good, if not 100% guaranteed way of ensuring the right codebase is being used. We can make it less likely that sidekiq loads in different versions of files to what's loaded in the test environment by moving the `start_sidekiq` call out of the `before each` and into a `before suite` block, but then would lose the modularity of only starting it when needed.

If your concern is simply that this shouldn't be in the scope of a _system_ test, then I don't see a problem with moving this to an _integration_ or _feature_ test, if you feel those describe what's being tested here better.

Reply:
That's a very good point about potential inconsistency between the code Sidekiq would be running versus that being tested, if the Sidekiq process was pre-run. So, yes, let's have this helper start and stop Sidekiq as you advise.

However, it does still replicate functionality supplied by the Sidekiq shell scripts and API in terms of spawning processes, stopping them, and checking for their existence.

By contrast, letting Sidekiq do most of the work for us:
```ruby
# frozen_string_literal: true

require 'sidekiq/api'

PID_FILE = File.join(Rails.root, 'tmp', 'pids', 'sidekiq.test.pid')
LOG_FILE = File.join(Rails.root, 'log', 'sidekiq.test.log')

def read_pid
  return nil unless File.exists?(PID_FILE)
  pid = File.read(PID_FILE).strip
  pid.to_i unless pid.blank?
end

def bundle_exec(cmd)
  system("RAILS_ENV=#{ENV['RAILS_ENV']} bundle exec #{cmd}")
end

def start_sidekiq
  bundle_exec("sidekiq -L #{LOG_FILE} -P #{PID_FILE} -d")
end

def stop_sidekiq
  bundle_exec("sidekiqctl stop #{PID_FILE}")
end

def sidekiq_running?
  Sidekiq::ProcessSet.new.any? { |ps| ps['pid'] == read_pid }
end

RSpec.configure do |config|
  config.before(:each) do |example|
    if example.metadata[:sidekiq]
      start_sidekiq unless sidekiq_running?
      Sidekiq::Testing.disable!
    end
  end

  config.after(:each) do |example|
    Sidekiq::Testing.fake! if example.metadata[:sidekiq]
  end

  config.after(:suite) do
    stop_sidekiq
  end
end
```

Reply:
yes that's good. I've tried various cases of interrupting tests and having other sidekiq processes running and this seems resilient to those sort of scenarios.